warp/yul/DeadcodeEliminator.py

```diff
@@ -0,0 +1,171 @@
+from functools import reduce
+from typing import List, Sequence, Set, Tuple, Union
+
+import yul.yul_ast as ast
+from yul.AstMapper import AstMapper
+
+LiveVars = Set[str]
+
+
+class DeadcodeEliminator(AstMapper):
+    def __init__(self):
+        super().__init__()
+
+    def map(self, node: ast.Node, **kwargs) -> ast.Node:
+        (vs, node) = self.visit(node)
+        assert (
+            vs == set()
+        ), "There should be no uninstantiated variables in the top scope."
+        return node
+
+    def visit_list(
+        self, nodes: Sequence[ast.Node], *args, **kwargs
+    ) -> (LiveVars, List[ast.Node]):
+        if len(nodes) == 0:
+            return (set(), list())
+        lis = [self.visit(x, *args, **kwargs) for x in nodes]
+        [var_set, nodes] = list(map(list, zip(*lis)))
+        live_vars = reduce(lambda s1, s2: s1 | s2, var_set, set())
+        return (live_vars, nodes)
+
+    def visit_literal(self, node: ast.Literal) -> Tuple[LiveVars, ast.Literal]:
+        return (set(), node)
+
+    def visit_identifier(self, node: ast.Identifier) -> Tuple[LiveVars, ast.Identifier]:
+        return (set([node.name]), node)
+
+    def visit_typed_name(self, node: ast.TypedName):
+        return (set([node.name]), node)
+
+    def visit_function_call(
+        self, node: ast.FunctionCall
+    ) -> (LiveVars, ast.FunctionCall):
+        (live_vars, arguments) = self.visit(node.arguments)
+        return (
+            live_vars,
+            ast.FunctionCall(function_name=node.function_name, arguments=arguments),
+        )
+
+    def visit_expression_statement(
+        self, node: ast.ExpressionStatement
+    ) -> (LiveVars, ast.ExpressionStatement):
+        (live_vars, _) = self.visit(node.expression)
+        return (live_vars, node)
+
+    def visit_assignment(self, node: ast.Assignment) -> Tuple[LiveVars, ast.Assignment]:
+        (live_vars, _) = self.visit(node.value)
+        return (live_vars, node)
+
+    def visit_variable_declaration(
+        self, node: ast.VariableDeclaration
+    ) -> (LiveVars, ast.VariableDeclaration):
+        if node.value == None:
+            return (set(), node)
+        (live_vars, _) = self.visit(node.value)
+        return (live_vars, node)
+
+    def visit_function_definition(
+        self, node: ast.FunctionDefinition
+    ) -> (LiveVars, ast.FunctionDefinition):
+        # We need to ignore the generated getters from kudu since their
+        # body is only instantiated in ToCairoVisitor
+        if node.name.startswith("getter"):
+            return (set(), node)
+        (live_vars, body) = self.visit(
+            node.body, return_vars=set([n.name for n in node.return_variables])
+        )
+        live_vars = live_vars - set([arg.name for arg in node.parameters])
+        # The warp ifs are the only place where the if branch has a return in each branch
```

Comment:
Not really, we don't extract into `__warp_if` if-expressions where at least one branch results in return